### PR TITLE
Surface SSH reconnect lifecycle for Cloud VMs

### LINKED ISSUE #3776: Make Cloud VM SSH sessions resilient to sleep and reconnects
Cloud VM dogfood surfaced that SSH sessions disconnect when the client machine sleeps or the network drops, and recovery is not clear enough.

Acceptance criteria:

- Cloud VM terminal panes surface a reconnecting state when the local connection drops.
- `cmux ssh` and `cmux vm ssh` reconnect automatically when possible.
- Existing remote process/session state remains usable after reconnect, or the UI clearly states that the remote session ended.
- Add a smoke test or documented manual smoke for sleep/network interruption against a Cloud VM.

Source: internal Austin/Konsti Cloud VM dogfood thread, May 8 2026.

### PR DESCRIPTION
**Pull request by lawrencecchen**
## Summary
- Reports SSH reconnecting and connected lifecycle from the foreground startup wrapper.
- Routes those lifecycle events through workspace remote state so Cloud VM panes show reconnecting and connected accurately.
- Keeps injected SSH LocalCommand helpers overridable by user-provided LocalCommand or PermitLocalCommand.

## Testing
- xcodebuild test -quiet -project GhosttyTabs.xcodeproj -scheme cmux-unit -configuration Debug -destination 'platform=macOS,arch=arm64' -derivedDataPath /tmp/cmux-issue-3776-tests -resultBundlePath /tmp/cmux-issue-3776-ssh-startup-suite.xcresult -only-testing:cmuxTests/CLINotifyProcessIntegrationRegressionTests/testSSHPaneCloseSignalDoesNotReportSessionEndToSharedTransport -only-testing:cmuxTests/CLINotifyProcessIntegrationRegressionTests/testSSHStartupRetriesTransientSSHExitBeforeReportingSessionEnd -only-testing:cmuxTests/CLINotifyProcessIntegrationRegressionTests/testSSHStartupReportsReconnectingAndConnectedLifecycleAroundTransientExit -only-testing:cmuxTests/CLINotifyProcessIntegrationRegressionTests/testSSHStartupStopsAtConfiguredReconnectLimit -only-testing:cmuxTests/CLINotifyProcessIntegrationRegressionTests/testSSHStartupDoesNotRetryNonTransientSSHExit -only-testing:cmuxTests/CLINotifyProcessIntegrationRegressionTests/testSSHSignalDerivedChildExitReportsSessionEnd -only-testing:cmuxTests/CLINotifyProcessIntegrationRegressionTests/testSSHSignalDuringReconnectDelayDoesNotStartAnotherSSH -only-testing:cmuxTests/CLINotifyProcessIntegrationRegressionTests/testSSHStartupPrintsFinalErrorBannerWhenStderrIsCaptured
- xcodebuild test -quiet -project GhosttyTabs.xcodeproj -scheme cmux-unit -configuration Debug -destination 'platform=macOS,arch=arm64' -derivedDataPath /tmp/cmux-issue-3776-tests -resultBundlePath /tmp/cmux-issue-3776-workspace.xcresult -only-testing:cmuxTests/WorkspaceRemoteConnectionTests/testRemoteTerminalLifecycleEventsDriveReconnectState
- python3 -m json.tool Resources/Localizable.xcstrings
- git diff --check
- ./scripts/reload.sh --tag issue-3776-ssh

## Issues
- Closes https://github.com/manaflow-ai/cmux/issues/3776

---

> [!NOTE]
> **Medium Risk**
> Medium risk because it changes SSH startup command generation and remote-connection state transitions, which could impact connection reliability and user-visible status during SSH sessions.
> 
> **Overview**
> Surfaces **SSH session lifecycle** into the app by adding internal CLI subcommands `ssh-session-reconnecting` and `ssh-session-connected`, emitting them from the SSH startup wrapper (and `vm ssh-attach`) during transient exit/retry and on successful connection.
> 
> Wires new RPCs `workspace.remote.terminal_reconnecting`/`terminal_connected` through `TerminalController` into `Workspace`, which now tracks pending lifecycle events per surface/`relay_port`, updates remote state to *reconnecting* with localized detail, and only flips to *connected* once proxy/daemon readiness (or VM no-proxy) conditions are met.
> 
> Tightens LocalCommand injection behavior by only installing helpers when safe (`LocalCommand` not already set and `PermitLocalCommand` allows it), threads `--relay-port` through split attach, adds new localizations and CLI contract docs, and expands unit/integration tests around these edge cases and validations.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 9456c492df24d762d2e880bcc8edcff40ca77f0e. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

---
## Summary by cubic
Cloud VM panes now use the native reconnect lifecycle from `CloudTuiManualMirrorSession` and workspace remote-terminal liveness, addressing #3776 without restoring obsolete SSH `LocalCommand` status callbacks.

- Preserves per-attempt ownership, pending readiness handling, and rejection of events for ended sessions.
- The merged tree matches `main`; no SSH configuration or migration changes are required.

<sup>Written for commit 75220e30daa883cd3df7aa38681b6fc1c631ee7e. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/manaflow-ai/cmux/pull/3930?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

* **New Features**
  * Remote terminal now reports reconnecting and connected lifecycle states (includes attempt/exit-status and relay-port handling). VM attach accepts a relay-port and can include local connected/timing scripts when allowed.

* **Bug Fixes / Reliability**
  * More reliable application/retry of pending "connected" updates and cleanup of ended sessions to reduce stale states.

* **Localization**
  * Added localized status messages for remote terminal connection states.

* **Tests**
  * Expanded lifecycle and security tests covering reconnect, connected, transient exits, and edge cases.

* **Documentation**
  * CLI lifecycle commands documented.

[Review Change Stack](https://app.coderabbit.ai/change-stack/manaflow-ai/cmux/pull/3930)

### CODE REVIEW

File: CLI/cmux.swift [hunk 1]
```diff
@@ -4833,12 +4838,21 @@ struct CMUXCLI {
...
-            relayPort: sshOptions.remoteRelayPort
-        )
+        let sshSessionConnectedCommandScript = canInstallLocalCommand
+            ? sshSessionConnectedLocalCommandScript(
+                remoteRelayPort: sshOptions.remoteRelayPort,
+                localCLIPath: resolvedExecutableURL()?.path
+            )
+            : nil
+        let sshConnectionTimingCommandScript = canInstallLocalCommand
+            ? sshConnectionTimingLocalCommandScript(
```
[the remark this answers is not in the document]
Comment (lawrencecchen) [reply CLI/cmux.swift:4858]: Fixed in d6884b90a. Reconnecting lifecycle is now gated by the same connected-lifecycle availability, so we do not enter reconnecting when LocalCommand cannot provide the connected callback.

*— Claude Code*

File: Sources/Workspace.swift [hunk 2]
```diff
@@ -9358,6 +9358,61 @@ final class Workspace: Identifiable, ObservableObject {
...
+
+    private func remoteTerminalLifecycleMatches(surfaceId: UUID, relayPort: Int?) -> Bool {
+        guard let relayPort,
+              relayPort > 0,
+              remoteConfiguration?.relayPort == relayPort else {
+            return false
+        }
+        return activeRemoteTerminalSurfaceIds.contains(surfaceId)
+            || pendingRemoteTerminalChildExitSurfaceIds.contains(surfaceId)
+    }
```
[the remark this answers is not in the document]
Comment (lawrencecchen) [reply Sources/Workspace.swift:9591]: Fixed in d6884b90a. Lifecycle matching now only accepts active remote terminal surfaces, and a connected-then-ended pre-config race is tombstoned so it cannot publish connected after teardown.

*— Claude Code*

File: docs/cli-contract.md [hunk 3]
```diff
@@ -153,6 +153,8 @@ Environment:
 | `vm-ssh-attach` | Hidden compatibility alias for older VM workspaces. |
 | `vm-pty-connect` | Internal helper that connects to a VM PTY from a config file. |
 | `ssh-session-end` | Internal helper that clears remote SSH session state. |
+| `ssh-session-reconnecting` | Internal helper that marks an SSH-backed workspace as reconnecting. |
+| `ssh-session-connected` | Internal helper that marks an SSH-backed workspace as connected after retry. |
```
[the remark this answers is not in the document]
Comment (lawrencecchen) [reply docs/cli-contract.md]: Fixed in d6884b90a. The CLI contract now says connected after reconnect for consistent terminology.

*— Claude Code*

**Comment by teamleaderleo** [comment]
cmux-reconcile: close-candidate

**Proposed action:** Close this empty PR without merging; preserve the branch.

**Evidence checked September 18, 2026:** GitHub reports **0 changed files, 0 additions, and 0 deletions**. I independently fetched the PR diff and it is empty. Head: `75220e30daa883cd3df7aa38681b6fc1c631ee7e`. Target branch: `main`.

There is no remaining patch in this PR against its target branch. This does **not** establish that the original feature shipped to `main` or that its idea should be discarded. If this is an intentional placeholder for future work, keep the underlying request and record that purpose; otherwise it is ready for queue cleanup.

Recheck the head/diff before acting in case new work arrives. Search `"cmux-reconcile: close-candidate"` in this repository for this cleanup batch.

**Comment by lawrencecchen** [comment]
Fleet instruction update for head `75220e30daa883cd3df7aa38681b6fc1c631ee7e`: this PR is classified `other`. No macOS build tag is claimed. The current controller app recipe does not establish iOS/test readiness; that requires the appropriate validated recipe. Use `cmux-ci` for supported jobs, retain the returned ID and receipt, and wait on the same ID after any timeout. Do not use retired maclease allocation or post credentials. Exact-head tags will be posted only after the applicable build succeeds.

**Comment by teamleaderleo** [comment]
Closing as already on main: merging this branch into main at 8421357dfc produces main's own tree, so there's nothing left to land. The branch is kept; reopen if something here is still missing. Part of the backlog cleanup in manaflow-ai/cmuxterm-hq#563.
